### PR TITLE
[Compression] update distillation utils

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
This pr mainly about offline distil utils.

ref: [Offline Distillation Dataloader Generation Utils](https://github.com/microsoft/nni/wiki/%5B20221114%5D-Distillation-Design#offline-distillation-dataloader-generation-utils)

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/contribute/distillation/uid_dataset.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        cpu_rng_state = torch.random.get_rng_state()
+
+        if self._replay_mode:
+            seed = self._suid_seed[str(suid)].pop(-1)
+            self._suid_seed[str(suid)].insert(0, seed)
+        else:
+            seed = self._generate_seed()
+            self._suid_seed[str(suid)].append(seed)
+        torch.random.manual_seed(seed)
+        sample = self._transform(sample)
```
Comment (wkcn) [inline nni/contribute/distillation/uid_dataset.py:96]: The function `self._transform` may use other random libraries, like `np.random` and `builtin.random`. We need to save and restore these states.

File: nni/contribute/distillation/__init__.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (matluster) [inline nni/contribute/distillation/__init__.py:1]: nni/contrib/distillation

File: nni/contribute/distillation/preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+from torch.utils.data import DataLoader, Dataset
+
+from .storage import create_storage
+from .uid_dataset import create_uid_dataset, _UidDataset
+from .utils import _default_labels_split_fn, _default_labels_collate_fn
+
+_logger = logging.getLogger(__name__)
+
+
+class Preprocessor:
```
Comment (matluster) [inline nni/contribute/distillation/preprocessor.py:21]: Is this ready for review?

Comment (J-shang) [reply nni/contribute/distillation/preprocessor.py:21]: yes, please review except `HDF5Storage`

File: nni/contribute/distillation/uid_dataset.py (hunk 1 above)
Comment (J-shang) [reply nni/contribute/distillation/uid_dataset.py]: I add `np.random`, `builtin.random`, `torch.random` rng save/load as default, do you think here need to give user an interface to customize save/load for some special random libraries we don't know?

Comment (wkcn) [reply nni/contribute/distillation/uid_dataset.py]: Great! Yes, it is better to provide an interface to customize the function `_default_manual_seed`.

File: nni/contrib/distillation/preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    def _patched_uid_dataloader(self):
+        dataloader = self._create_dataloader(self._dataset)
+        origin_collate_fn = dataloader.collate_fn
+
+        def collate_fn(data):
+            uids, origin_data = list(zip(*data))
+            return uids, origin_collate_fn(origin_data)  # type: ignore
+
+        setattr(dataloader, 'collate_fn', collate_fn)
```
Comment (matluster) [inline nni/contrib/distillation/preprocessor.py:88]: What's wrong with `dataloader.collate_fn = colalte_fn`?

File: nni/contrib/distillation/preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        `IndexedDataset` will be used to wrap `dataset`. `IndexedDataset` take the index passed to `Dataset.__getitem__` as data's uid.
+        `uid_dataset_cls` should be a subclass of `_UidDataset`, `uid_dataset_cls.__getitem__` will return a tuple of (uid, original_data).
+        (Official uid dataset and how to customize uid dataset, please refer ...)
+    uidd_args
+        Additional arguments pass to `uid_dataset_cls.__init__`.
+        If there are many extra arguments, recommended to directly set `uid_dataset_cls` to None, and then pass in the wrapped `dataset`.
+    uidd_kwargs
+        Additional keyword arguments pass to `uid_dataset_cls.__init__`.
+        If there are many extra arguments, recommended to directly set `uid_dataset_cls` to None, and then pass in the wrapped `dataset`.
+    labels_split_fn
```
Comment (matluster) [inline nni/contrib/distillation/preprocessor.py:57]: So the "label" here is not label, but rather distillation signals.
Is that a commonly-used term?

File: nni/contrib/distillation/uid_dataset.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    def __init__(self, dataset: Dataset, *args, **kwargs):
+        assert not isinstance(dataset, IterableDataset), 'Do not support `IterableDataset`.'
+        self._dataset = dataset
+        self._replay_mode = False
+
+    def __len__(self):
+        return len(self._dataset)  # type: ignore
+
+    def __getitem__(self, index):
+        return None, self._dataset.__getitem__(index)
```
Comment (matluster) [inline nni/contrib/distillation/uid_dataset.py:25]: NotImplementedError?

File: nni/contrib/distillation/__init__.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/contrib/distillation/__init__.py:1]: `Preprocessor`, `Preprocessor.preprocess_labels` and `Preprocessor.create_replay_dataloader` is the first level API, in most cases, calling these three is enough.

File: nni/contrib/distillation/preprocessor.py (hunk 4 above)
Comment (J-shang) [reply nni/contrib/distillation/preprocessor.py:88]: nothing wrong, just a preference

File: nni/contrib/distillation/uid_dataset.py (hunk 6 above)
Comment (J-shang) [reply nni/contrib/distillation/uid_dataset.py:25]: yes, `NotImplementedError` is better

File: nni/contribute/distillation/__init__.py (hunk 2 above)
Comment (matluster) [reply nni/contribute/distillation/__init__.py]: On a second thought, `nni/contrib/distill` might be even better, because it's shorter.

But since `nni.compression` is already long, you can also keep the name `distillation` if you want to stick to the tradition.

File: nni/contrib/distillation/preprocessor.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        def collate_fn(data):
+            uids, origin_batch = uid_collate_fn(data)
+            distil_labels = list(map(self._storage.select, cast(List[str], uids)))
+            distil_labels = labels_collate_fn(distil_labels)
+            return distil_labels, origin_batch
+
+        setattr(dataloader, 'collate_fn', collate_fn)
+        return dataloader
+
+    def preprocess_labels(self, epochs: int = 1, log_process: bool = True, labels_split_fn: Callable[[Any], List] | None = None):
```
Comment (matluster) [inline nni/contrib/distillation/preprocessor.py:105]: Did you discuss about renaming this?

Comment (J-shang) [reply nni/contrib/distillation/preprocessor.py:105]: `generate_distillation_labels`? I will rename it.
Do you think cls name `Preprocessor` should be renamed? Maybe `DistillationLabelGenerator`?

File: nni/contrib/distillation/uid_dataset.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+    def __getitem__(self, index):
+        raise NotImplementedError()
+
+    def get_origin_dataset(self):
+        return self._dataset
+
+    def observe(self):
+        """
+        Observe mode means this dataset is on processing by `Preprocessor`
```
Comment (Bonytu) [inline nni/contrib/distillation/uid_dataset.py:32]: when will observe() be used?

Comment (J-shang) [reply nni/contrib/distillation/uid_dataset.py:32]: If user need more distillation labels by run `preprocess_labels` second time, he needs to use `observe()` to set `Preprocessor` to observe mode.

File: nni/contrib/distillation/storage.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+    def cleanup(self):
+        """
+        Cleanup the saved files under `cache_folder`.
+        """
+        cleanup_filename = set(self._uid2filename.values())
+        if Path(self._cache_folder / SETTING_FILENAME).exists():
+            with Path(self._cache_folder / SETTING_FILENAME).open(mode='r') as f:
+                uid2filename: Dict[str, str] = json_tricks.load(f)
+                cleanup_filename = cleanup_filename.difference(uid2filename.values())
```
Comment (Bonytu) [inline nni/contrib/distillation/storage.py:155]: according to this description, cleanup function is used to cleanup the saved files under cache_folder,  why dose use cleanup_filename,difference() to find cleanup_filename?  When will cleanup_filename and uid2filename be different?

Comment (J-shang) [reply nni/contrib/distillation/storage.py:155]: it's a good question. you can find `if Path(self._cache_folder / SETTING_FILENAME).exists()` is True, that means this is a checkpoint folder, and some filenames are loaded from the `Path(self._cache_folder / SETTING_FILENAME)`, then when cleanup, we should only delete the files generated by this storage, not the files loaded from the checkpoint.

File: nni/contrib/distillation/preprocessor.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        self._dataset.observe()
+
+        dataloader = self._patched_uid_dataloader()
+        total_batch_num = len(dataloader)
+        log_literal = min(max(total_batch_num // 25, 10), 100)
+
+        for epoch_idx in range(epochs):
+            for batch_idx, packed_batch in enumerate(dataloader):
+                uids, batch = packed_batch
+                soft_labels = self._teacher_predict(batch)
```
Comment (Bonytu) [inline nni/contrib/distillation/preprocessor.py:141]: maybe teacher_model is better?

File: nni/contrib/distillation/preprocessor.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+                soft_labels = labels_split_fn(soft_labels)
+
+                assert isinstance(soft_labels, abc.Iterable)
+                list(map(self._storage.record, uids, soft_labels))
+                if log_process and ((batch_idx + 1) % log_literal == 0 or (batch_idx + 1) == total_batch_num):
```
Comment (Bonytu) [inline nni/contrib/distillation/preprocessor.py:147]: maybe we can rename log_literal

File: nni/contrib/distillation/preprocessor.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        self._storage = create_storage(keep_in_memory=keep_in_memory, cache_folder=cache_folder, cache_mode=cache_mode,
+                                       checkpoint_folder=checkpoint_folder)
+
+        self._teacher_predict = teacher_predict
+        self._dataset = create_uid_dataset(dataset, uid_dataset_cls, uidd_args=uidd_args, uidd_kwargs=uidd_kwargs)
+        self._create_dataloader = create_dataloader
+
+        assert (labels_split_fn is None and labels_collate_fn is None) or (labels_split_fn is not None and labels_collate_fn is not None)
+        self._labels_split_fn = labels_split_fn if labels_split_fn is not None else _default_labels_split_fn
+        self._labels_collate_fn = labels_collate_fn if labels_collate_fn is not None else _default_labels_collate_fn
```
Comment (Bonytu) [inline nni/contrib/distillation/preprocessor.py:76]: when will the self._labels_collate_fn be used?

File: nni/contrib/distillation/preprocessor.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        The iteration of dataloader will return (distil_labels, origin_batch).
+
+        Parameters
+        ----------
+        labels_collate_fn
+            Use to create replayed dataloader.
+            A function used to collate the samples from distillation storage. If `labels_collate_fn` is None,
+            the samples are excepted to be tensors and have same size, they will be batched and return.
+        """
+        labels_collate_fn = labels_collate_fn if labels_collate_fn is not None else _default_labels_collate_fn
```
Comment (Bonytu) [inline nni/contrib/distillation/preprocessor.py:167]: why not use self._labels_collate_fn?

File: nni/contrib/distillation/preprocessor.py (hunk 11 above)
Comment (J-shang) [reply nni/contrib/distillation/preprocessor.py:141]: in fact, here we need a predict function, not a model

File: nni/contrib/distillation/preprocessor.py (hunk 12 above)
Comment (J-shang) [reply nni/contrib/distillation/preprocessor.py:147]: good suggestion

File: nni/contrib/distillation/preprocessor.py (hunk 14 above)
Comment (J-shang) [reply nni/contrib/distillation/preprocessor.py:167]: nice catch, it should be `self._labels_collate_fn`

File: nni/contrib/distillation/preprocessor.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+                list(map(self._storage.record, uids, soft_labels))
+                if log_process and ((batch_idx + 1) % log_literal == 0 or (batch_idx + 1) == total_batch_num):
+                    info_msg = f'[epoch {epoch_idx}: {batch_idx + 1} / {total_batch_num}]'
+                    _logger.info(info_msg)
+
+        # ready for creating replay dataloader
+        self._preprocessed = True
+        self._dataset.replay()
+
+    def create_replay_dataloader(self, labels_collate_fn: Callable[[List], Any] | None = None):
```
Comment (Bonytu) [inline nni/contrib/distillation/preprocessor.py:155]: it seems that I feel confused about replay_dataloader unless I read this code. So is it necessary to rename replay_dataloader?

Comment (J-shang) [reply nni/contrib/distillation/preprocessor.py:155]: any suggestion?

Comment (Bonytu) [reply nni/contrib/distillation/preprocessor.py:155]: I have no good idea yet, maybe reuse_dataloader?

**Review by Bonytu — APPROVED** [review]

File: nni/contrib/distillation/label_patcher.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+from torch.utils.data import DataLoader, Dataset
+
+from .storage import create_storage
+from .uid_dataset import create_uid_dataset, _UidDataset
+from .utils import _default_labels_split_fn, _default_labels_collate_fn
+
+_logger = logging.getLogger(__name__)
+
+
+class DistilLabelPatcher:
```
Comment (matluster) [inline nni/contrib/distillation/label_patcher.py:21]: Distill has double-l.

Comment (J-shang) [reply nni/contrib/distillation/label_patcher.py:21]: distil and distill are both ok, and in distillation, distil seems more normal, i.e. DistilBert, DistilGBT2
